src/main/java/ui/components/pickers/LabelPickerState.java

```diff
@@ -0,0 +1,272 @@
+        if (!hasExactlyOneMatchedLabel(repoLabels, keyword)) return this;
+        String labelName = getMatchedLabelName(repoLabels, keyword);
+
+        if (isAnInitialLabel(labelName)) {
+            if (isARemovedLabel(labelName)) {
+                // add back initial label
+                removeConflictingLabels(labelName);
+                removedLabels.remove(labelName);
+            } else {
+                removedLabels.add(labelName);
+            }
+        } else {
+            if (isAnAddedLabel(labelName)) {
+                addedLabels.remove(labelName);
+            } else {
+                // add new label
+                removeConflictingLabels(labelName);
+                addedLabels.add(labelName);
+            }
+        }
+
+        return new LabelPickerState(initialLabels, addedLabels, removedLabels, new ArrayList<>());
+    }
+
+    public LabelPickerState nextSuggestion() {
+        if (canIncreaseIndex()) {
+            return new LabelPickerState(initialLabels, removedLabels, addedLabels,
+                    matchedLabels, OptionalInt.of(currentSuggestionIndex.getAsInt() + 1));
+        }
+        return this;
+    }
+
+    public LabelPickerState previousSuggestion() {
+        if (canDecreaseIndex()) {
+            return new LabelPickerState(initialLabels, removedLabels, addedLabels,
+                        matchedLabels, OptionalInt.of(currentSuggestionIndex.getAsInt() - 1));
+        }
+        return this;
+    }
+
+    public LabelPickerState updateMatchedLabels(Set<String> repoLabels, String query) {
+        List<String> newMatchedLabels = convertToList(repoLabels);
+
+        newMatchedLabels = filterByName(newMatchedLabels, getName(query));
+        newMatchedLabels = filterByGroup(newMatchedLabels, getGroup(query));
+
+        OptionalInt newSuggestionIndex;
+        if (query.equals("") || newMatchedLabels.isEmpty()) {
+            newSuggestionIndex = OptionalInt.empty();
+        } else {
+            newSuggestionIndex = OptionalInt.of(0);
+        }
+
+        return new LabelPickerState(initialLabels, addedLabels, removedLabels, newMatchedLabels, newSuggestionIndex);
+    }
+
+    /*
+     * Methods for updating UI
+     */
+
+    public List<String> getAssignedLabels() {
+        List<String> assignedLabels = new ArrayList<>();
+        assignedLabels.addAll(initialLabels);
+        assignedLabels.addAll(addedLabels);
+        assignedLabels.removeAll(removedLabels);
+        return assignedLabels;
+    }
+
+    public List<String> getInitialLabels() {
+        return convertToList(initialLabels);
+    }
+
+    public List<String> getRemovedLabels() {
+        return convertToList(removedLabels);
+    }
+
+    public List<String> getAddedLabels() {
+        return convertToList(addedLabels);
+    }
+
+    public Optional<String> getCurrentSuggestion() {
+        if (currentSuggestionIndex.isPresent() && isValidSuggestionIndex()) {
```

Comment:
`isValidSuggestionIndex()` should probably be checked in an assertion, since it's a programmer error if it returns false.